### PR TITLE
feat: upgrade docker image to py38 and add support for py39

### PR DESCRIPTION
**Pull request by villebro**
### SUMMARY
With Python 3.10 right around the corner (2021-10-04) and multiple performance improvements when going from 3.7 to 3.8 (see Python Speed Center benchmark summary below), we should move to 3.8 and start preparing for transitioning to 3.9. This PR bumps the official image from Python 3.7 to 3.8 and adds support for Python 3.9. CI is also bumped from 3.7 to 3.8 (btw, `tox` was already confgured with `basepython = python3.8`), and dual tests for 3.7 and 3.8 are replaced with 3.8 and 3.9.

### Performance benchmark
On most benchmark tests, 3.8 (blue) performs slightly better than 3.7 (orange):
screencapture-speed-python-org-comparison-2021-09-29-10_32_18

### TESTING INSTRUCTIONS
Test locally that everything works as expected

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: requirements/base.txt [hunk 1]
```diff
@@ -257,7 +257,7 @@ sqlalchemy==1.3.24
     #   flask-sqlalchemy
     #   marshmallow-sqlalchemy
     #   sqlalchemy-utils
-sqlalchemy-utils==0.36.8
+sqlalchemy-utils==0.37.8
```
Comment (villebro) [inline requirements/base.txt:260]: `sqlalchemy-utils==0.36.8` has been yanked from PyPI, causing an ugly warning when installing deps. In addition, `py39` was added to the officially supported versions right after `0.37.8` was cut (https://github.com/kvesteri/sqlalchemy-utils/commit/b5b48b13a22aa6d796e98c0225f367a2f010c788), so I assume `0.37.8` to be fully supported by 3.9 in practice.

File: requirements/development.in [hunk 2]
```diff
@@ -18,7 +18,7 @@
 -r base.in
 flask-cors>=2.0.0
 mysqlclient==1.4.2.post1
-pillow>=7.0.0,<8.0.0
+pillow>=8.3.1,<9
```
Comment (villebro) [inline requirements/development.in:21]: Pillow 7.x caused problems on Python 3.9

File: superset/utils/memoized.py [hunk 3]
```diff
@@ -64,7 +64,9 @@ def __get__(
         if not self.is_method:
             self.is_method = True
         # Support instance methods.
-        return functools.partial(self.__call__, obj)
+        func = functools.partial(self.__call__, obj)
+        func.__func__ = self.func  # type: ignore
+        return func
```
Comment (villebro) [inline superset/utils/memoized.py:69]: I wasn't quite able to figure out why this is necessary, but running doctests on memoized methods started failing on 3.9 due to the 
The only thing I found is this which seems like the same problem, but is said to affect Python 3.8, not 3.9: https://bugs.python.org/issue12790 Hacky workaround, but does the trick. I'll try to figure out how to solve this more elegantly before this gets merged

File: setup.py [hunk 4]
```diff
@@ -168,5 +168,6 @@ def get_git_sha() -> str:
     classifiers=[
         "Programming Language :: Python :: 3.7",
```
Comment (amitmiran137) [inline setup.py:169]:
```suggestion
```

**Review by amitmiran137 — APPROVED** [review]
LGTM
Minor change is needed

File: setup.py (hunk 4 above)
Comment (etr2460) [reply setup.py:169]: I think we're still going to support 3.7 (we support last 3 python versions officially). Once 3.10 is out, we'll remove 3.7 support in a major version bump

Comment (villebro) [reply setup.py:169]: I want to deprecate 3.7 too, but I think we should give it some more time (it's still currently the recommended version) - let's leave it alive for a month or so so everyone can get upgraded to at least 3.8

Comment (john-bodley) [reply setup.py:169]: @etr2460 @villebro do we—Apache Superset—have an official policy about supporting the last three Python versions? Just asking. If that's the case I think this is great as it means we're not unnecessarily burdened by older less feature rich versions.

Comment (dpgaspar) [reply setup.py:169]: @john-bodley @etr2460 I think we just lightly agreed on the last 3 versions on a community meeting. Making it an official policy is a good idea to let users know what to expect and plan ahead

Comment (villebro) [reply setup.py:169]: @john-bodley TBH until this PR we were only really supporting two 😄 But yeah, I think there was an unwritten agreement to try to support 3. I'm fine cutting it down to "at least 2" but trying to support 3 if possible.

File: superset/utils/memoized.py (hunk 3 above)
Comment (dpgaspar) [reply superset/utils/memoized.py:69]: weird!

**Review by dpgaspar — APPROVED** [review]

**Comment by villebro** [comment]
Opened a task on Apache Infra to bump required tests: https://issues.apache.org/jira/browse/INFRA-22377

**Pull request merged**